applications/nrf9160dk_demo/src/main.c

```diff
@@ -0,0 +1,1029 @@
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_TOPIC_MSG
+	};
+	size_t len = strlen(modem_at_cmd_buff);
+
+	if (len == 0) {
+		err = -ENOBUFS;
+		state = AT_CMD_ERROR;
+	} else {
+#if defined(CONFIG_AT_CMD)
+		err = at_cmd_write(modem_at_cmd_buff, modem_at_cmd_buff,
+				sizeof(modem_at_cmd_buff), &state);
+#else
+		/* Proper error msg has already been copied to buffer */
+		err = 0;
+#endif
+	}
+
+	/* Get response length; same buffer was used for the response. */
+	len = strlen(modem_at_cmd_buff);
+
+	if (err) {
+		len = snprintf(modem_at_cmd_buff, sizeof(modem_at_cmd_buff),
+				"AT CMD error: %d, state %d", err, state);
+	} else if (len == 0) {
+		len = snprintf(modem_at_cmd_buff, sizeof(modem_at_cmd_buff),
+				"OK\r\n");
+	} else if (len > MODEM_AT_CMD_MAX_RESPONSE_LEN) {
+		len = snprintf(modem_at_cmd_buff, sizeof(modem_at_cmd_buff),
+				MODEM_AT_CMD_RESP_TOO_BIG_STR);
+	}
+
+	modem_data.data.buf = modem_at_cmd_buff;
+	modem_data.data.len = len;
+
+	err = cloud_encode_data(&modem_data, CLOUD_CMD_GROUP_COMMAND, &msg);
+	if (err) {
+		LOG_ERR("[%s:%d] cloud_encode_data failed with error %d",
+			__func__, __LINE__, err);
+	} else {
+		err = cloud_send(cloud_backend, &msg);
+		cloud_release_data(&msg);
+		if (err) {
+			LOG_ERR("[%s:%d] cloud_send failed with error %d",
+				__func__, __LINE__, err);
+		}
+	}
+
+	k_sem_give(&modem_at_cmd_sem);
+}
+
+/**@brief Send button presses to cloud */
+static void button_send(u8_t button_num, bool pressed)
+{
+	static char data[3];
+
+	if (!atomic_get(&send_data_enable)) {
+		return;
+	}
+
+	data[0] = button_num + '0';
+	if (pressed) {
+		data[1] = '1';
+	} else {
+		data[1] = '0';
+	}
+	data[2] = '\0';
+
+	button_cloud_data.data.buf = data;
+	button_cloud_data.data.len = strlen(data);
+	button_cloud_data.tag += 1;
+
+	if (button_cloud_data.tag == 0) {
+		button_cloud_data.tag = 0x1;
+	}
+
+	LOG_INF("Sending button event for button %d=%s", button_num,
+		pressed ? "pressed" : "released");
+	k_work_submit_to_queue(&application_work_q, &send_button_data_work);
+
+	if ((button_num == 4) && pressed) {
+		send_signon_message();
+	}
+}
+
+/**@brief Send a text message to the cloud; work function
+ * will need to free the duplicated string; duplication is done
+ * in case message is in a temporary buffer
+ */
+static void msg_send(const char *message)
+{
+	if (!atomic_get(&send_data_enable)) {
+		return;
+	}
+
+	msg_cloud_data.data.buf = strdup(message);
+	msg_cloud_data.data.len = strlen(message);
+	msg_cloud_data.tag += 1;
+
+	if (msg_cloud_data.tag == 0) {
+		msg_cloud_data.tag = 0x1;
+	}
+
+	LOG_INF("Sending message: %s", message);
+	k_work_submit_to_queue(&application_work_q, &send_msg_data_work);
+}
+
+static void cloud_cmd_handle_modem_at_cmd(const char * const at_cmd)
+{
+	if (!at_cmd) {
+		return;
+	}
+
+	if (k_sem_take(&modem_at_cmd_sem, K_MSEC(20)) != 0) {
+		LOG_ERR("[%s:%d] Modem AT cmd in progress.", __func__,
+		       __LINE__);
+		return;
+	}
+
+#if defined(CONFIG_AT_CMD)
```

Comment:
can we remove most of these `#if`s?  (AT, modem config, bsd lib, fota, dk lib, mcu boot)
this demo app should always be built with these options enabled (maybe the default .conf needs updated).
i think it looks cleaner.

Reply:
yes, agreed

Reply:
Actually, Chris Hansen suggested that the remote AT command feature was unnecessary for this example, so it should be removed, not enabled by default as you suggest.  What do you think?

Reply:
sure, if you are removing the feature no need to enable it by default.